kani-compiler/src/kani_middle/coercion.rs

```diff
@@ -0,0 +1,188 @@
+//!
+//! The more complicated coercions are DST Coercions (aka Unsized Coercions). These coercions
+//! allow rust to create references to dynamically sized types (such as Traits and Slices) by
+//! casting concrete sized types. Unsized coercions can also be used to cast unsized to unsized
+//! types. These casts work not only on the top of references, but it also handle
+//! references inside structures, allowing the unsized coercions of smart pointers. This last
+//! piece was defined by
+//! [RFC 982 DST Coercion](https://rust-lang.github.io/rfcs/0982-dst-coercion.html).
+
+use rustc_hir::lang_items::LangItem;
+use rustc_middle::traits::{ImplSource, ImplSourceUserDefinedData};
+use rustc_middle::ty::adjustment::CustomCoerceUnsized;
+use rustc_middle::ty::TypeAndMut;
+use rustc_middle::ty::{self, ParamEnv, TraitRef, Ty, TyCtxt};
+use rustc_span::symbol::Symbol;
+use tracing::trace;
+
+/// Extract the pair (`T`, `U`) for a unsized coercion where type `T` implements `Unsize<U>`.
+/// I.e., `U` is either a trait or a slice.
+/// For more details, please refer to:
+/// <https://doc.rust-lang.org/reference/type-coercions.html#unsized-coercions>
+///
+/// This is used to determine the vtable implementation that must be tracked by the fat pointer.
+///
+/// For example, if `&u8` is being converted to `&dyn Debug`, this method would return:
+/// `(u8, dyn Debug)`.
+///
+/// There are a few interesting cases (references / pointers are handled the same way):
+/// 1. Coercion between `&T` to `&U`.
+///    - This is the base case.
+///    - In this case, we extract the type that is pointed to.
+/// 2. Coercion between smart pointers like `Rc<T>` to `Rc<U>`.
+///    - Smart pointers implement the `CoerceUnsize` trait.
+///    - Use CustomCoerceUnsized information to traverse the smart pointer structure and find the
+///      underlying pointer.
+///    - Use base case to extract `T` and `U`.
+/// 3. Coercion between `&Wrapper<T>` to `&Wrapper<U>`.
+///    - Apply base case to extract the pair `(Wrapper<T>, Wrapper<U>)`.
+///    - Extract the tail element of the struct which are of type `T` and `U`, respectively.
+/// 4. Coercion between smart pointers of wrapper structs.
+///    - Apply the logic from item 2 then item 3.
+pub fn extract_unsize_casting<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    src_ty: Ty<'tcx>,
+    dst_ty: Ty<'tcx>,
+) -> (Ty<'tcx>, Ty<'tcx>) {
+    trace!(?dst_ty, ?src_ty, "find_trait_conversion");
+    let info = CoerceUnsizedIterator::new(tcx, src_ty, dst_ty).last().unwrap();
+    let src_base_ty = extract_pointee(info.src_ty)
+        .expect(&format!("Expected source to be a pointer. Found {:?} instead", info.src_ty));
+    let dst_base_ty = extract_pointee(info.dst_ty)
+        .expect(&format!("Expected destination to be a pointer. Found {:?} instead", info.dst_ty));
+    tcx.struct_lockstep_tails_erasing_lifetimes(src_base_ty, dst_base_ty, ParamEnv::reveal_all())
```

Comment:
You described 4 cases, but all the magic happens here?

Reply:
Not really. Lines 55 / 57 would be enough for case 1. Line 54 is needed for case 2 and 4. This logic is needed for cases 3 and 4. I.e., this takes care of traversing the concrete type after we traversed the pointer type.

Reply:
I added some comments so hopefully it helps understand the logic.